kaldb/src/main/java/com/slack/kaldb/chunkManager/ChunkCleanerService.java

```diff
@@ -1,54 +1,89 @@
 package com.slack.kaldb.chunkManager;
 
-import com.google.common.annotations.VisibleForTesting;
-import com.google.common.util.concurrent.AbstractScheduledService;
 import com.slack.kaldb.chunk.Chunk;
 import com.slack.kaldb.chunk.ChunkInfo;
 import java.time.Duration;
 import java.time.Instant;
 import java.util.ArrayList;
+import java.util.Comparator;
 import java.util.List;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /**
  * The chunk cleaner service removes snapshot chunks after a configurable duration from the chunk
- * manager.
+ * manager and when the number of chunks goes over a certain amount.
  */
-public class ChunkCleanerService<T> extends AbstractScheduledService {
+public class ChunkCleanerService<T> {
 
   private static final Logger LOG = LoggerFactory.getLogger(ChunkCleanerService.class);
 
+  private final ChunkManager<T> chunkManager;
+  private final int limit;
+
   private final Duration staleDelayDuration;
-  private final IndexingChunkManager chunkManager;
 
-  public ChunkCleanerService(IndexingChunkManager chunkManager, Duration staleDelayDuration) {
+  public ChunkCleanerService(ChunkManager<T> chunkManager, int limit, Duration staleDelayDuration) {
```

Comment:
I'm not sure I like the design of this in light of the new changes we've made. Have we considered getting rid of this `ChunkCleanerService` class entirely, and moving these stale chunks to the indexer implementations specifically?

The idea that we need to keep multiple stale ones only applies to the indexer anyhow - we're just setting it to "0" for the recovery nodes to reuse the code.

It seems like we could move all this complexity internal to the indexing chunk manager, and then the recovery indexer has a much simpler cleanup logic. Then we also don't expose the extra methods to the chunk manager interface.

Reply:
I think that removing the abstraction would be a good idea here. The problem is that for one context (indexer) we have the idea of a limit, delay duration - and the only other context we invoke this method (recovery) we just pass empty values (ie `new ChunkCleaner<>(this, 0, Duration.ZERO);`). 

This also helps solves the other issue of needing to expose these to the `ChunkManager` interface and then implementing them with `UnsupportedOperationException` in the cache. You wouldn't need to add any methods to the interface then, and the the behavior is sufficiently different between the recovery and index nodes that it'll make it easier to read if they're separate implementations.